injector/__init__.py

```diff
@@ -784,10 +797,25 @@ def get(self, key: Type[T], provider: Provider[T]) -> Provider[T]:
         try:
             return self._context[key]
         except KeyError:
-            provider = InstanceProvider(provider.get(self.injector))
+            instance = self._get_instance(key, provider, self.injector)
+            provider = InstanceProvider(instance)
             self._context[key] = provider
             return provider
 
+    def _get_instance(self, key: Type[T], provider: Provider[T], injector: 'Injector') -> T:
+        if injector.parent and not injector.binder.has_explicit_binding_for(key):
+            try:
+                return self._get_instance_from_parent(key, provider, injector.parent)
+            except (CallError, UnsatisfiedRequirement):
+                pass
+        return provider.get(injector)
+
+    def _get_instance_from_parent(self, key: Type[T], provider: Provider[T], parent: 'Injector') -> T:
+        singleton_scope_binding, _ = parent.binder.get_binding(type(self))
+        singleton_scope = singleton_scope_binding.provider.get(parent)
+        provider = singleton_scope.get(key, provider)
+        return provider.get(parent)
```

Comment:
At first I duplicated more of `Injector.get()`, like this:

```suggestion
        binding, binder = parent.binder.get_binding(key)

        scope = binding.scope or type(self)  # TODO(dp): Unsure how to do this best
        if isinstance(scope, ScopeDecorator):
            scope = scope.scope
        singleton_scope_binding, _ = binder.get_binding(scope)
        scope_instance = singleton_scope_binding.provider.get(self)

        provider = scope_instance.get(key, provider)
        return provider.get(parent)
```

It has more complexity, and I could not create a test case that proved its necessity. Can you think of one?